### PR TITLE
fix a bug that spring-mvc set an error endpoint name 

### PR DESCRIPTION
**Pull request by darcydai**
fix a bug that spring-mvc set an error endpoint name if the controller class annotation implements an interface.

----------------------------
recently, I find some error logs in the OAP server like this 
image

after decoding the endpoint id, I found the endpoint name is empty. after adding more logs in the trace segment receiver, I found the error from the spring-MVC plugin. 

 if the controller class annotation implements an interface, it can not get the RequestMapping. in the spring-cloud  tech stack, it is common sense, like RPC framework is feign
## example
``` java
@RequestMapping(value = "/for")
public interface A {
    @RequestMapping
    String bar();
}
```

```java
@RestController
public class B implements A{

    @Override
    public String bar() {
        return "oops";
    }
}
```  
  
  
result in oap server :
image

### Fix <spring-mvc set an error endpoint name if the controller class annotation implements an interface>
- [ ] Add a unit test to verify that the fix works.
- [x] Explain briefly why the bug exists and how to fix it.

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

**Comment by ascrutae** [comment]
You'd better add this case into the test scenarios about Spring framework.  @darcydai

**Comment by ascrutae** [comment]
> You'd better add this case into the test scenarios about Spring framework. @darcydai

Here are those test scenarios about Spring framework:
1. [spring-4.3.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.3.x-scenario)
2. [spring-3.1.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-3.1.x-scenario)
3. [spring-4.1.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.1.x-scenario)
4. [spring-4.3.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.3.x-scenario)

**Comment by darcydai** [comment]
> > You'd better add this case into the test scenarios about Spring framework. @darcydai
> 
> Here are those test scenarios about Spring framework:
> 1. [spring-4.3.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.3.x-scenario)
> 2. [spring-3.1.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-3.1.x-scenario)
> 3. [spring-4.1.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.1.x-scenario)
> 4. [spring-4.3.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.3.x-scenario)

I will do this later

**Comment by darcydai** [comment]
> > You'd better add this case into the test scenarios about Spring framework. @darcydai
> 
> Here are those test scenarios about Spring framework:
> 
> 1. [spring-4.3.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.3.x-scenario)
> 2. [spring-3.1.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-3.1.x-scenario)
> 3. [spring-4.1.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.1.x-scenario)
> 4. [spring-4.3.x-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/spring-4.3.x-scenario)
I found  a scenarios can only define one entry service,can i add more scenario in spring mvc for the special case ?  
@ascrutae  
image

**Comment by wu-sheng** [comment]
> I found a scenarios can only define one entry service,can i add more scenario in spring mvc for the special case ?

You don't need to. Do a cycle call inside the entrance service. You could have as many `/PATH` as you like, unless timeout due to too slow

**Comment by darcydai** [comment]
> > I found a scenarios can only define one entry service,can i add more scenario in spring mvc for the special case ?
> 
> You don't need to. Do a cycle call inside the entrance service. You could have as many `/PATH` as you like, unless timeout due to too slow

thanks, i got it

**Comment by darcydai** [comment]
I found a new bug when I coding the test case for spring MVC, I has fix it in last commit.

**Comment by darcydai** [comment]
DO NOT MERGE UNTIL I FINISH SCENARIOS TEST CASE

**Comment by darcydai** [comment]
I found a new bug when i coding the scenarios for the spring MVC plugin. i have fixed in last commit
 `org.apache.skywalking.apm.plugin.spring.mvc.commons.interceptor.RequestMappingMethodInterceptor#getRequestURL`
1. the annotation RequestMapping with the value[value()] array length equals zore represent an empty string in spring MVC concept.
image

2. RequestMapping in spring mvc 3.x and 4.1  has no the path method [path()], if RequestMapping with  with the value[value()] array length equals zore in spring mvc 3 will case no such method error;
image

image


3. because of AliasFor annotation, no need to call path(), I have tested
image


image

**Comment by darcydai** [comment]
do not  merge before i finish scenario test for spring mvc.

**Comment by wu-sheng** [comment]
> do not merge before I finish the scenario test for spring MVC.

 I expect you could update the plugin test scenarios.

**Comment by wu-sheng** [comment]
@darcydai Please fix your test case.

**Comment by wu-sheng** [comment]
Seems still have style issues?

**Comment by darcydai** [comment]
> Seems still have style issues?

seems failed scenario test is not spring MVC, how to trigger  the task run again

**Comment by wu-sheng** [comment]
> > Seems still have style issues?
> 
> seems failed scenario test is not spring MVC, how to trigger the task run again

I will try to rerun, I think we have located the issue.

File: test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml [hunk 1]
```diff
@@ -14,52 +14,130 @@
...
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 14
+            isError: false
+            spanType: Entry
+            peer: ''
+            skipAnalysis: false
+            tags:
+              - {key: url, value: 'http://localhost:8080/spring-3.0.x-scenario//impl/requestmapping'}
```
Comment (ascrutae) [inline test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml:52]: The URL which has more '/' char looks strange, I think that `http://localhost:8080/spring-3.0.x-scenario/impl/requestmapping` prefer looks good.

File: test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml [hunk 2]
```diff
@@ -14,52 +14,130 @@
...
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 14
+            isError: false
+            spanType: Entry
+            peer: ''
+            skipAnalysis: false
+            tags:
+              - {key: url, value: 'http://localhost:8080/spring-3.0.x-scenario//impl/restmapping'}
```
Comment (ascrutae) [inline test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml:74]: Same as above

File: test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml [hunk 3]
```diff
@@ -14,52 +14,130 @@
...
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 13
+            isError: false
+            spanType: Exit
+            peer: localhost:8080
+            skipAnalysis: false
+            tags:
+              - {key: url, value: 'http://localhost:8080/spring-3.0.x-scenario//impl/requestmapping'}
```
Comment (ascrutae) [inline test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml:108]: Same as above

File: test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml [hunk 4]
```diff
@@ -14,52 +14,130 @@
...
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 13
+            isError: false
+            spanType: Exit
+            peer: localhost:8080
+            skipAnalysis: false
+            tags:
+              - {key: url, value: 'http://localhost:8080/spring-3.0.x-scenario//impl/restmapping'}
```
Comment (ascrutae) [inline test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml:123]: Same as above

File: test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml (hunk 1 above)
Comment (ascrutae) [reply test/plugin/scenarios/spring-3.0.x-scenario/config/expectedData.yaml:52]: I found that this problem caused by `TestServiceBean#URL` has more  '/',   You can remove it to resolve the problem

**Review by ascrutae — APPROVED** [review]
LGTM

**Pull request merged**